### DISCUSSION TITLE
singularity: unable to execute code

### DISCUSSION THREAD

**Discussion by erikajob91** [firedrakeproject/firedrake#2816] [Firedrake support]
I opened  the firedrakeproject/firedrake:latest docker image via singularity on my HPC cluster and I got this situation:

_### Singularity> ls -alrt
total 1
drwxr-xr-x 3 root        root      32 Feb 19 02:31 .
drwxr-x--- 6 root        root     216 Feb 19 03:13 firedrake
drwxr-xr-x 1 epellegrino calcoli3  60 Mar 13 11:18 .._

**_Singularity> sudo firedrake
sudo: The "no new privileges" flag is set, which prevents sudo from running as root.
sudo: If sudo is running in a container, you may need to adjust the container configuration to disable the flag._**

Actually I'm not able to copy the directory firedrake either.

**Comment by tranner** [comment]
You need to activate the virtual environment first and then you can run the python code in that environment. Singularity doesn't do that for you. You may also need to map some directories from singularities rootless point of view:

```
mkdir -p .run .cache
$ singularity shell -B .run:/run/user -B .cache:/home/firedrake/firedrake/.cache firedrake.sif # enter singularity container
Apptainer> . /home/firedrake/firedrake/bin/activate # activate the virtual environment
(firedrake) Apptainer> python ..... # run your code!
```
